src/fifo.c

```diff
@@ -0,0 +1,305 @@
+            fifoBlock *newblock = zmalloc(sizeof(fifoBlock));
+            newblock->u.last_or_first_idx = 0;
+            newblock->items[0] = ptr;
+            q->last->u.next = newblock; /* overwrites the index, setting it to 0 */
+            q->last = newblock;
+        }
+    }
+
+    q->length++;
+}
+
+
+/* Look at the first item in the queue (without removing it).
+ * NOTE: asserts if the queue is empty. */
+void *fifoPeek(fifo *q) {
+    assert(q->length > 0);
+    int firstIdx = (q->first == q->last) ? 0 : q->first->u.last_or_first_idx & IDX_MASK;
+    return q->first->items[firstIdx];
+}
+
+
+/* Return and remove the first item from the queue.
+ * NOTE: asserts if the queue is empty. */
+void *fifoPop(fifo *q) {
+    assert(q->length > 0);
+    void *item;
+
+    if (q->first == q->last) {
+        /* With only 1 block, POP occurs at index 0 and items 1..6 are shifted. */
+        item = q->last->items[0];
+
+        int lastIdx = q->last->u.last_or_first_idx; /* pointer portion is 0 on last (or only) block */
+        assert(lastIdx < ITEMS_PER_BLOCK);
+
+        if (lastIdx > 0) {
+            /* With only 1 block, shift the items rather than eventually needing new block.
+             * (This is cheap, shifting a max of 6 pointers.) */
+            for (int i = 0; i < lastIdx; i++) q->last->items[i] = q->last->items[i + 1];
+            q->last->u.last_or_first_idx--; /* Decrement the last index */
+        } else {
+            /* Just finished the only block.  Delete it. */
+            zfree(q->last);
+            q->first = q->last = NULL;
+        }
+    } else {
+        /* With more than 1 block, POP occurs at firstIdx, and firstIdx is incremented. */
+        int firstIdx = q->first->u.last_or_first_idx & IDX_MASK;
+        item = q->first->items[firstIdx];
+
+        if (firstIdx < ITEMS_PER_BLOCK - 1) {
+            /* Just increment the first index to the next slot. */
+            q->first->u.last_or_first_idx++;
+        } else {
+            /* Finished with this block, move to next */
+            q->first->u.last_or_first_idx &= ~IDX_MASK; /* restores the next pointer */
+            fifoBlock *next = q->first->u.next;
+            zfree(q->first);
+            q->first = next;
+        }
+    }
+
+    q->length--;
+
+    return item;
+}
+
+
+/* Return the number of items in the queue. */
+long fifoLength(fifo *q) {
+    return q->length;
+}
+
+
+/* Delete the queue.
+ * NOTE: this does not free items which may be referenced by inserted pointers. */
+void fifoDelete(fifo *q) {
+    if (q->length > 0) {
+        fifoBlock *cur = q->first;
+        while (cur != NULL) {
+            cur->u.last_or_first_idx &= ~IDX_MASK; /* zero out the last 3 bits */
+            fifoBlock *next = cur->u.next;
+            zfree(cur);
+            cur = next;
+        }
+    }
+    zfree(q);
+}
+
+
+/* Blindly overwrites target from source. */
+static void blindlyMoveFifoContents(fifo *target, fifo *source) {
```

Comment:
```suggestion
/* Overwrites target from source. */
static void overwriteFifoContents(fifo *target, fifo *source) {
```
Blindly is a little confusing. Would rather use more inclusive language.